editor/js/editor-libs/telemetry.js

```diff
@@ -0,0 +1,92 @@
+import { postParentMessage } from "./events.js";
+import { getStorageItem, storeItem } from "./utils.js";
+
+const ACTION_COUNTS_KEY = "action-counts";
+
+/**
+ * Reads action counts from local storage.
+ * Ignores action counts that belong to another example.
+ *
+ * @returns The action counts from the previous session, if available.
+ */
+function getActionsCounts() {
+  const json = getStorageItem(ACTION_COUNTS_KEY);
+
+  if (!json) {
+    return {};
+  }
+
+  const value = JSON.parse(json);
+  if (value && value.href === window.location.href && value.counts) {
+    return value.counts;
+  }
+
+  return {};
+}
+
+/**
+ * Writes action counts to local storage, to persist them between reloads.
+ *
+ * @param {object} counts - The current action counts.
+ */
+function storeActionCounts(counts) {
+  storeItem(
+    ACTION_COUNTS_KEY,
+    JSON.stringify({
+      href: window.location.href,
+      counts,
+    })
+  );
+}
+
+/**
+ * Action counts by key.
+ * Used to distinguish 1st/2nd/etc occurrences of the same event.
+ */
+const actionCounts = getActionsCounts();
+
+/**
+ * Last observed action.
+ * Used to ignore multiple input actions until another action happened.
+ */
+let lastAction = null;
+
+/**
+ * Records an action, by counting it and forwarding it to the window parent.
+ *
+ * @param {string} key - Distinct name of the type of action.
+ * @param {boolean} deduplicate - Should multiple actions of the same type be ignored until another action occurred?
+ */
+export function recordAction(key, deduplicate = false) {
+  actionCounts[key] = actionCounts[key] ?? 0;
+
+  if (deduplicate && key === lastAction) {
+    return;
+  } else {
+    lastAction = key;
+  }
+
+  const source = `${key} -> ${actionCounts[key]}`;
+  actionCounts[key]++;
+  storeActionCounts(actionCounts);
+
+  postParentMessage("action", { source });
+}
+
+document.addEventListener("DOMContentLoaded", () => {
```

Comment:
I think it would be better to replace it with a `register` or `registerTelemetry` function. Generally, modules in the project don't have side effects and always some function is called instead. For example `mceConsole` in `console.js`, `register` in `events.js`, `initEditor` and `registerEventListeners` in `tabby.js`

Reply:
Resolved via b931bdd7844b8b53ee2b2add910330cd3ac5e23b.